ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaMain.cpp

```diff
@@ -21,37 +21,56 @@
 
 #include <unistd.h>
 #include <stdio.h>
+#include <mach-o/dyld.h>
+#include <limits.h>
 #include "CocoaMain.h"
 #include "ObjectiveC.h"
 #include <cstdlib>
+#include <string>
 
 #include "../General/PFwindow.h"
 #include "../General/PFfilemanip.h"
 
 #include "Platforms/General/PFmain.h"
 #include "Universal_System/roomsystem.h"
 
+using std::string;
+
 namespace enigma {
   void SetResizeFptr();
-}
+  
+  void initialize_directory_globals() {
+    // Set the working_directory
+    char buffer[PATH_MAX + 1];
+    if (getcwd(buffer, PATH_MAX + 1) != NULL)
+      enigma_user::working_directory = buffer;
 
-namespace enigma_user {
-  std::string working_directory = "";
+    // Set the program_directory
+    buffer[0] = 0;
+    char real_executable[PATH_MAX + 1];
+    char *bundle_id;
+
+    uint32_t bufsize = sizeof(buffer);
+    if (_NSGetExecutablePath(buffer, &bufsize) == 0) {
+      bundle_id = dirname(buffer);
+      strcpy(real_executable, bundle_id);
+      strcat(real_executable, "/");
```

Comment:
Note that this is vastly less efficient than just using a `std::string` and reserve. Each time you run strcat, you are reading the entire string. It's also incredibly ugly, so I'm not sure why you keep using this C pattern in C++ code.

Reply:
resolved.